internal/db/embedding.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2025 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/philippgille/chromem-go"
+
+	"github.com/sourcenetwork/defradb/client"
+	"github.com/sourcenetwork/defradb/errors"
+	"github.com/sourcenetwork/defradb/internal/keys"
+)
+
+func getEmbeddingFunc(provider, model, url string) (chromem.EmbeddingFunc, error) {
+	switch provider {
+	case "ollama":
+		return chromem.NewEmbeddingFuncOllama(model, url), nil
+	case "openai":
+		normalized := true
+		apiURL := url
+		if apiURL == "" {
+			apiURL = chromem.BaseURLOpenAI
+		}
+		return chromem.NewEmbeddingFuncOpenAICompat(
+			apiURL,
+			os.Getenv("OPENAI_API_KEY"),
+			model,
+			&normalized,
+		), nil
+	default:
+		return nil, NewErrUnknownEmbeddingProvider(provider)
+	}
+}
+
+// setEmbedding sets the embedding fields on the document if the related fields are dirty.
+func (c *collection) setEmbedding(ctx context.Context, doc *client.Document, isCreate bool) error {
+	embeddingGenerated := false
+	for _, colField := range c.Definition().GetFields() {
+		if colField.Embedding != nil {
+			fieldsVal := make(map[string]client.NormalValue)
+			needsGeneration := false
```

Comment:
question: It looks like if a user explicitly provides their own vector value it will be overwritten by the result of the remote call.  Would it not be better to allow users to provide their own values?

Besides being less surprising for them, it would give them greater control over the determinism of the resultant CID/DocID, which I believe can vary depending on the timing of the remote calls (under most AI models).

Reply:
It shouldn't overwrite. If it does it's a mistake on my part. Users should be free to use their own vectors.

Reply:
Ah okay nice!  

todo: I've not read the tests yet, could you please make sure that there is at least one that covers that scenario?

suggestion: Please also note this behaviour in the func documentation

Reply:
Done and done